### PR TITLE
gRPC FaaS PoC

### PR DESCRIPTION
**Pull request by tjholm**

### CODE REVIEW

File: faas/faas.go [hunk 1]
```diff
@@ -15,29 +15,117 @@
...
+	"google.golang.org/grpc"
 )
 
 // NitricFunction - a function built using Nitric, to be executed
-type NitricFunction func(*NitricRequest) *NitricResponse
+type NitricFunction func(*NitricTrigger) (*NitricResponse, error)
+
+func faasLoop(stream pb.Faas_TriggerStreamClient, f NitricFunction, errorCh chan error) {
+	for {
+		// Block recieving a message
```
Comment (jyecusch) [inline faas/faas.go:31]: *receiving*

File: faas/faas.go [hunk 2]
```diff
@@ -15,29 +15,117 @@
...
+			break
+		}
+
+		// We have a trigger
+		if srvrMsg.GetTriggerRequest() != nil {
+			req, err := FromGrpcTriggerRequest(srvrMsg.GetTriggerRequest())
+
+			if err != nil {
+				fmt.Println("There was an error reading the TriggerRequest", err)
+				// Return a bad request here...
```
Comment (jyecusch) [inline faas/faas.go:49]: should this be returning the error response?

File: faas/faas.go [hunk 3]
```diff
@@ -15,29 +15,117 @@
...
+
+				continue
+			}
+			// Let the membrane know the function is ready for initializatio
```
Comment (jyecusch) [inline faas/faas.go:53]: Should this comment still be here? If yes, it has a typo at the end.

File: faas/faas.go [hunk 4]
```diff
@@ -15,29 +15,117 @@
...
+			// Process this trigger
+			response, err := f(req)
+
+			if err != nil {
+				fmt.Println("Function return an error", err)
```
Comment (jyecusch) [inline faas/faas.go:58]: _returned_

File: faas/response.go [hunk 5]
```diff
@@ -15,22 +15,46 @@
...
-		ctx.Response.Header.Add(k, v)
+func (n *NitricResponse) SetData(data []byte) {
+	n.data = data
+}
+
+func (n *NitricResponse) GetContext() *ResponseContext {
+	return n.context
+}
+
+// ToTriggerResponse - Tranlates a Nitric Response for gRPC transport to the memrbane
```
Comment (jyecusch) [inline faas/response.go:35]: *membrane*

File: faas/trigger.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	data     []byte
+	mimeType string
+}
+
+// GetContext - return the context of a request, with metadata about that request.
+func (n *NitricTrigger) GetContext() *NitricTriggerContext {
+	return n.context
+}
+
+// GetPayload - return the []byte payload of the request.
```
Comment (jyecusch) [inline faas/trigger.go:35]: *GetData*

File: faas/trigger.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+func (n *NitricTrigger) GetData() []byte {
+	return n.data
+}
+
+func (n *NitricTrigger) GetMimeType() string {
```
Comment (jyecusch) [inline faas/trigger.go:40]: Missing doc comment

File: faas/trigger.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+	return n.mimeType
+}
+
+// GetStruct - Unmarshals the request body from JSON to the provided interface{}
```
Comment (jyecusch) [inline faas/trigger.go:44]: *GetDataAsStruct*

File: faas/trigger.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+func (n *NitricTrigger) GetDataAsStruct(object interface{}) error {
+	return json.Unmarshal(n.data, object)
+}
+
+// Default Response - Returns a default response object dependent on the Trigger context
```
Comment (jyecusch) [inline faas/trigger.go:49]: *DefaultResponse*

File: faas/trigger.go [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+func (n *NitricTrigger) DefaultResponse() *NitricResponse {
+
+	var context interface{} = nil
+
+	if n.context.IsHttp() {
+		context = &HttpResponseContext{
+			Headers: make(map[string]string),
+			Status:  200,
+		}
+	} else if n.context.IsHttp() {
```
Comment (jyecusch) [inline faas/trigger.go:59]: The else condition is identical to the if condition. I think it should be 'IsTopic'. Probably need a test for this.

File: faas/trigger.go [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+	return &NitricResponse{
+		data: nil,
+		context: &ResponseContext{
+			context: context,
+		},
+	}
+}
+
+// fromHttpRequest - converts a standard nitric HTTP request into a NitricRequest to be passed to functions.
```
Comment (jyecusch) [inline faas/trigger.go:73]: *FromGrpcTriggerRequest* - converts a gRPC TriggerRequest into a NitricTrigger to be passed to a function.

File: faas/trigger_context.go [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+import (
+	pb "github.com/nitrictech/go-sdk/interfaces/nitric/v1"
+)
+
+// NitricTriggerContext - Represents the contextual metadata for a Nitric function request.
+type NitricTriggerContext struct {
+	context interface{}
+}
+
+func (c *NitricTriggerContext) IsHttp() bool {
```
Comment (jyecusch) [inline faas/trigger_context.go:26]: *IsHttp* - Returns true if the trigger was via an HTTP Request and HTTP metadata is available

File: faas/trigger_context.go [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+	_, ok := c.context.(*NitricHttpTriggerContext)
+
+	return ok
+}
+
+func (c *NitricTriggerContext) IsTopic() bool {
```
Comment (jyecusch) [inline faas/trigger_context.go:32]: IsTopic - Returns true if the trigger was an Event via a Topic and Event/Topic metadata is available

File: faas/trigger_context.go [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+	_, ok := c.context.(*NitricTopicTriggerContext)
+
+	return ok
+}
+
+func (c *NitricTriggerContext) AsHttp() *NitricHttpTriggerContext {
```
Comment (jyecusch) [inline faas/trigger_context.go:38]: Missing doc comment

File: faas/trigger_context.go [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+	if ctx, ok := c.context.(*NitricHttpTriggerContext); ok {
+		return ctx
+	}
+
+	return nil
+}
+
+func (c *NitricTriggerContext) AsTopic() *NitricTopicTriggerContext {
```
Comment (jyecusch) [inline faas/trigger_context.go:46]: Missing doc comment

File: faas/trigger_context.go [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+	Headers     map[string]string
+	Path        string
+	QueryParams map[string]string
+}
+
+type NitricTopicTriggerContext struct {
+	Topic string
+}
+
+func ContextFromTriggerRequest(grpcTrigger *pb.TriggerRequest) *NitricTriggerContext {
```
Comment (jyecusch) [inline faas/trigger_context.go:65]: Missing doc comment

**Review by jyecusch — CHANGES_REQUESTED** [review]

File: faas/faas.go (hunk 1 above)
Comment (tjholm) [reply faas/faas.go:31]: Fixed

File: faas/faas.go (hunk 2 above)
Comment (tjholm) [reply faas/faas.go:49]: Fixed.

File: faas/faas.go (hunk 3 above)
Comment (tjholm) [reply faas/faas.go:53]: Fixed.

**Review by jyecusch — DISMISSED** [review]

**Review by jyecusch — APPROVED** [review]

**Pull request merged**